controllers/dockermachine_controller.go

```diff
@@ -91,137 +132,138 @@ func (r *DockerMachineReconciler) Reconcile(req ctrl.Request) (_ ctrl.Result, re
+		if err := externalMachine.CreateControlPlane(machine.Spec.Version); err != nil {
+			return ctrl.Result{}, errors.Wrap(err, "failed to create control plane DockerMachine")
+		}
+	} else {
+		if err := externalMachine.CreateWorker(machine.Spec.Version); err != nil {
+			return ctrl.Result{}, errors.Wrap(err, "failed to create worker DockerMachine")
+		}
+	}
 
-// String is a helper function to print a human-readable form of the state.
-func (s State) String() string {
-	switch s {
-	case Provisioning:
-		return "Provisioning"
-	case Deleted:
-		return "Deleted"
-	case Provisioned:
-		return "Provisioned"
-	case Pending:
-		return "Pending"
-	default:
-		return "Unknown"
+	// if the machine is the first control plane added to the cluster, update the load balancer configuration
+	// NB. in case of the first control plane node this step must be executed before kubeadm init
+	if isFirstControlPlane {
+		if err := externalLoadBalancer.UpdateConfiguration(); err != nil {
+			return ctrl.Result{}, errors.Wrap(err, "failed to update DockerCluster.loadbalancer configuration")
+		}
 	}
-}
 
-func getState(machine *clusterv1.Machine, dockerMachine *infrav1.DockerMachine) State {
-	// Deleted takes precedence
-	if !machine.ObjectMeta.DeletionTimestamp.IsZero() {
-		return Deleted
+	// exec bootstrap
+	// NB. this step is necessary to mimic the behaviour of cloud-init that is embedded in the base images
+	// for other cloud providers
+	if err := externalMachine.ExecBootstrap(*machine.Spec.Bootstrap.Data); err != nil {
+		return ctrl.Result{}, errors.Wrap(err, "failed to exec DockerMachine bootstrap")
 	}
-	if dockerMachine.Spec.ProviderID != nil {
-		return Provisioned
+
+	// if the new machine is an additional control-plane node, update the load balancer configuration
+	// NB. in case of additional control plane nodes this step is executed after kubeadm join in order
+	// to not redirect traffic to an API server instance not yet ready.
+	if util.IsControlPlaneMachine(machine) && !isFirstControlPlane {
+		if err := externalLoadBalancer.UpdateConfiguration(); err != nil {
+			return ctrl.Result{}, errors.Wrap(err, "failed to update DockerCluster.loadbalancer configuration")
+		}
 	}
-	if machine.Spec.Bootstrap.Data == nil {
-		return Pending
+
+	// Set the provider ID on the Kubernetes node corresponding to the external machine
+	// NB. this step is necessary because there is no a cloudcontroller for docker that executes this step
+	if err := externalMachine.SetNodeProviderID(); err != nil {
+		return ctrl.Result{}, errors.Wrap(err, "failed to patch the Kubernetes node with the machine providerID")
 	}
-	return Provisioning
-}
 
-func (r *DockerMachineReconciler) create(
-	ctx context.Context,
-	c *clusterv1.Cluster,
-	machine *clusterv1.Machine,
-	dockerMachine *infrav1.DockerMachine) (ctrl.Result, error) {
+	// Set ProviderID so the Cluster API Machine Controller can pull it
+	providerID := externalMachine.ProviderID()
```

Comment:
No need to set the providerID on the Machine, it will be updated with the Provider ID from the DockerMachine in the capi Machine controller.

Reply:
@detiber I'm not sure I fully understand this comment... 
here we are setting the providerID on the DockerMachine (not on the machine)

Reply:
Ah, the naming `externalMachine` confused me on the last pass, thanks for the clarification.